### PR TITLE
Misc Job and N(ext) Runner (and other) fixes

### PR DESCRIPTION
**Pull request by clebergnu**
A collection of fixes and small improvements found while field testing the integrated (posting soon) integration of the N(ext) Runner and job.

### CODE REVIEW

File: avocado/core/task/statemachine.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ async def triage(self):
         except IndexError:
             return
 
-        requirements_ok = await self._spawner.spawn_task(runtime_task)
+        requirements_ok = await self._spawner.check_task_requirements(runtime_task)
         if requirements_ok:
```
Comment (pevogam) [inline avocado/core/task/statemachine.py:83]: Hi @clebergnu, good catch! I was just about to ask you about this and why we use the same status from `spawn_task` in both cases when we check for requirements and when we check if the task started successfully.

File: avocado/plugins/run.py [hunk 2]
```diff
@@ -253,7 +253,7 @@ def configure(self, parser):
                                  help_msg=help_msg,
                                  choices=('none', 'stdout', 'stderr',
                                           'both', 'combined', 'all'),
-                                 parser=parser,
+                                 parser=out_check,
                                  default=None,
```
Comment (pevogam) [inline avocado/plugins/run.py:257]: How is this change related to the current group of fixes? Do you mind elaborating with an extra sentence or two?

**Review by pevogam — COMMENTED** [review]
Hi @clebergnu, I have only looked at parts that I am currently dealing with but these are no doubt some good changes.

[*] Sometimes I wonder though why can't they make it in the original PR and have to be added in a follow up PR like this since the history could be more coherent if the branch introducing the changes also introduces the fixes and/or some extra testing. It would be good if in the future we work on slower but better quality PRs and are able to take our time to finish them completely and then proceed to merging but this is getting really off-topic for the current changes which although delayed are indeed good.

File: avocado/core/status/repo.py [hunk 3]
```diff
@@ -53,10 +59,25 @@ def get_latest_task_data(self, task_id):
...
+        status = message.get('status')
+        time = message.get('time')
+        if not all((task_id, status, time)):
+            return
+        if task_id not in self._status:
+            self._status[task_id] = (status, time)
+        else:
+            current_time = self._status[task_id][1]
+            if time > current_time:
+                self._status[task_id] = (status, time)
```
Comment (beraldoleal) [inline avocado/core/status/repo.py:74]: To avoid if else here, one suggestions is to have inside a try.. except:

```python
try:
    if time > self._status[task_id][1]:
        self._status[task_id] = (status, time)
except KeyError:
    self._status[task_id] = (status, time)
```

File: avocado/core/status/repo.py [hunk 4]
```diff
@@ -71,3 +92,8 @@ def process_raw_message(self, raw_message):
     @property
     def result_stats(self):
         return {key: len(value) for key, value in self._by_result.items()}
+
+    def get_task_status(self, task_id):
+        if task_id not in self._status:
+            return None
```
Comment (beraldoleal) [inline avocado/core/status/repo.py:98]: There is any reason to have this if, when you have the default values for the .get() bellow? Seems to me that only the return line would be enough here.

File: avocado/core/task/statemachine.py (hunk 1 above)
Comment (clebergnu) [reply avocado/core/task/statemachine.py:83]: And this was causing tests to run twice... which made me worry about the new code performance... finding the culprit saved my mood! :)

**Comment by clebergnu** [comment]
> Hi @clebergnu, I have only looked at parts that I am currently dealing with but these are no doubt some good changes.
> 
> [*] Sometimes I wonder though why can't they make it in the original PR and have to be added in a follow up PR like this since the history could be more coherent if the branch introducing the changes also introduces the fixes and/or some extra testing. It would be good if in the future we work on slower but better quality PRs and are able to take our time to finish them completely and then proceed to merging but this is getting really off-topic for the current changes which although delayed are indeed good.

Hi @pevogam, first of all thanks for the review.  Second, I hear you, I think a few of the commits here could have been held a bit more (until their users are also introduced)...

Yet, some of the changes are being added to a follow up PR because they were only identified/found after the original PR was out and even merged.

So to summarize it: point taken.  I'll try to avoid sending off-topic commits too early.

File: avocado/plugins/run.py (hunk 2 above)
Comment (clebergnu) [reply avocado/plugins/run.py:257]: Oh, this is just a usability fix related to Avocado jobs, that is, to `avocado run --help`.  It's not related to the nrunner though.

File: avocado/core/task/statemachine.py [hunk 5]
```diff
@@ -75,11 +75,11 @@ async def triage(self):
         """Reads from triaging, moves into either: ready or finished."""
         try:
             async with self._state_machine.lock:
-                runtime_task = self._state_machine.triaging.pop()
+                runtime_task = self._state_machine.triaging.pop(0)
         except IndexError:
             return
 
-        requirements_ok = await self._spawner.spawn_task(runtime_task)
+        requirements_ok = await self._spawner.check_task_requirements(runtime_task)
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/task/statemachine.py:82]: > I am curious how does this interplay with the `requirements` module (and resolver) overall? Is each spawner supposed to import and call its own resolver type (assuming different required backend)?

For now, each spawner implementation is canonical with regards to a task requirements/check (and in the future fulfillment).  The requirement themselves are coming from the `avocado.core.resolver` module, and so far, read them from `:avocado: requirement` docstring statements, but BP002 lists other ideas/possibilities.

> I think the current API seems to make the implicit assumption that a spawner's requirement resolution is tied together with its isolation model. What if for instance a podman spawner has to provide a specific container determined from a configuration?

So I envision that there may a different type of requirement, that may influence the spawner selection or actions.  For instance, one could say that test "foo" needs `vm.guest_os=fedora.32`, which will hardly be able to resolved by the "process" spawner.

To summarize it: for now, we only have functionality to deal with requirements *within* the isolation model, but something on top of it is thought to be a good idea and part of future development.

[the remark this answers is not in the document]
Comment (pevogam) [reply avocado/core/task/statemachine.py:82]: Alright, sounds great!

File: avocado/plugins/run.py (hunk 2 above)
Comment (pevogam) [reply avocado/plugins/run.py:257]: I see. My typical expectation when looking at a PR would be to have all changes in the file diff correspond to the topic of the PR though. I guess this is just a hit and miss situation.

File: avocado/core/status/repo.py (hunk 4 above)
Comment (clebergnu) [reply avocado/core/status/repo.py:98]: You're right, dropping this.

File: avocado/core/status/repo.py (hunk 3 above)
Comment (clebergnu) [reply avocado/core/status/repo.py:74]: It's more compact indeed... but I must say I find the readability a bit worse (my bias, of course).  Thanks for the suggestion, though.

**Comment by pevogam** [comment]
> > Hi @clebergnu, I have only looked at parts that I am currently dealing with but these are no doubt some good changes.
> > [*] Sometimes I wonder though why can't they make it in the original PR and have to be added in a follow up PR like this since the history could be more coherent if the branch introducing the changes also introduces the fixes and/or some extra testing. It would be good if in the future we work on slower but better quality PRs and are able to take our time to finish them completely and then proceed to merging but this is getting really off-topic for the current changes which although delayed are indeed good.
> 
> Hi @pevogam, first of all thanks for the review. Second, I hear you, I think a few of the commits here could have been held a bit more (until their users are also introduced)...
> 
> Yet, some of the changes are being added to a follow up PR because they were only identified/found after the original PR was out and even merged.
> 
> So to summarize it: point taken. I'll try to avoid sending off-topic commits too early.

I think my point in summary is simply that we should be able to take our time and propose a more final PRs with less pressure on merging, putting quality first. But I think we are already heading in this direction and are probably also waiting for the major 81-82 releases for things to calm down. Then PRs would probably have fewer follow-up PRs that simply finish the first PRs.

File: avocado/core/status/repo.py [hunk 6]
```diff
@@ -11,6 +11,12 @@ class StatusRepo:
     def __init__(self):
         #: Contains all reveived messages by a given task (by its ID)
         self._all_data = {}
+        #: Contains the most up to date status of a task, and the time it
+        #: it was set in a tuple (status, time).  This is keyed
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/status/repo.py:15]: Fixed, thanks!

File: avocado/core/status/repo.py [hunk 7]
```diff
@@ -71,3 +92,8 @@ def process_raw_message(self, raw_message):
     @property
     def result_stats(self):
         return {key: len(value) for key, value in self._by_result.items()}
+
+    def get_task_status(self, task_id):
+        if task_id not in self._status:
+            return None
+        return self._status.get(task_id, (None, None))[0]
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/status/repo.py:99]: Dirty little hack :)

File: avocado/core/status/repo.py [hunk 8]
```diff
@@ -53,10 +59,25 @@ def get_latest_task_data(self, task_id):
             return None
         return task_data[-1]
 
+    def _update_status(self, message):
+        """Update the latest status of atask (by time, not by message)."""
+        task_id = message.get('id')
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/status/repo.py:64]: Fixed, thanks!

**Comment by clebergnu** [comment]
Closing in favor of #4143
